### ISSUE TITLE
Deploy just one lambda function

### ISSUE THREAD

**Issue by sheerun** [danielcondemarin/serverless-next.js#185]
**Is your feature request related to a problem? Please describe.**

Current code deploys 2 lambda functions instead of 1, both of which need be kept warm which incurs additional charges.

**Describe the solution you'd like**

Would you accept pull request that makes this component deploy just one lambda function that is shared between api and normal pages?

**Comment by danielcondemarin** [comment]
@sheerun The decision to split the SSR and API pages into separate functions is very deliberate. 
The counter-argument to your premise is by splitting the functions you have smaller artifacts deployed onto each function which lead to better cold start times. 
Unless there is a big reason to do so, I prefer to keep the architecture design as is.

**Comment by sheerun** [comment]
Maybe would be best to allow both options and allow to split app into as many lambas as user wants (in my case just 1 is best). Something like:

```
myNextApp:
  component: serverless-next.js
  inputs:
    bucketName: my-bucket
    split:
      - /api
      - /admin
```

With configuration 3 lambdas would be created. One for all routes that start with /api, second for all routes that start with /admin and third for everything else.

**Comment by sheerun** [comment]
> which lead to better cold start times.

I'm pertty sure it's not the case becuse if front-end calls API, both of which are hosted in lambdas, then we need to wait one full cold time for front-end to start and then another one for api to start.

So it looks like this:

- request comes in
- front-end lambda cold starts (10s)
- front-end lambda calls API lambda to render
- API lambda cold starts  (10s)

So total cold time is 20s.

Compare to the case when everything is deployed in single lambda:

- request comes in
- front-end lambda cold starts (10s)
- front-end lambda calls API lambda to render
- API lambda is already hot  (0s)

Total cold request time is 0s.

Cold-start time is not very significantly dependent on lambda size especially if relevant codepath is loaded only on demand

**Comment by danielcondemarin** [comment]
> Maybe would be best to allow both options and allow to split app into as many lambas as user wants (in my case just 1 is best). Something like:
> 
> ```
> myNextApp:
>   component: serverless-next.js
>   inputs:
>     bucketName: my-bucket
>     split:
>       - /api
>       - /admin
> ```
> 
> With configuration 3 lambdas would be created. One for all routes that start with /api, second for all routes that start with /admin and third for everything else.

That would be nice but my concern is the complexity overhead it adds to the component for how much value?  Having the split between API and non-API seems like a good balance to me. 
Also, bear in mind CloudFront only allows [25 cache behaviours / paths to be defined](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/cloudfront-limits.html#limits-web-distributions). I'm already using 3 (`/static/*`, `/_next/*` and `/api/*`).

**Comment by sheerun** [comment]
Then at least please allow for deploying single lambda because of cold start issues I mentioned.

**Comment by eahefnawy** [comment]
Imo it’d be fine to have that as an option for the upsides that you mentioned. Just as long as we don’t compromise on the zero config design.

@danielcondemarin how complex do you think the implementation would be?

**Comment by danielcondemarin** [comment]
> Then at least please allow for deploying single lambda because of cold start issues I mentioned.

Just to clarify, I'm not entirely against this change. I'm simply trying to find more value in the change beyond just mitigating Cold Starts. From my personal experience I've not had 10+ seconds cold starts in Lambda@Edge or Lambda in general (as long as is outside a VPC).
I do appreciate different apps have different needs and access patterns, so an architecture where the user can determine how the Lambda functions are mapped to paths could be of benefit, with sensible defaults of course.  There is also an added benefit, users could deploy an individual part of the app. It could enable different teams working on the same application and having separate deployment workflows without interfering with each other.

> @danielcondemarin how complex do you think the implementation would be?

It depends on the approach. Deploying a single lambda function is probably not a lot of effort, that's actually how it used to work before I split it into 2 functions. But this would only cover one use case, mitigating the cold starts like @sheerun described above.

I'm more inclined towards the more flexible arch. 

```yaml
myNextApp:
  component: serverless-next.js
  inputs:
    bucketName: my-bucket
    split:
      - /api
      - /admin
      - /blog
      ...
```

This is maybe not a lot of work either, however I'm concerned about keeping track of all the Lambdas being provisioned. This is something you've raised before @eahefnawy as far as I can remember.
Also, there would be only up to 25 definable paths as that's the limit of how many Cache Behaviours CloudFront can have. 

I'll wait for a bit more feedback / conversation before deciding how to move forwards. I'm also interested on @ac360 opinion on this.

**Comment by ac360** [comment]
Before jumping into the solution, I'm curious about the underlying problem.  10-20s cold starts are very unusual (unless VPC is involved).

@sheerun could you provide more detail as to what's causing 10-20s cold starts?

**Comment by sheerun** [comment]
The point is not that it takes 20s but that it takes 2x longer and I also need to keep both lambdas hot

**Comment by sheerun** [comment]
To have total flexibility I'd suggest to allow each lambda to cover multiple routes, and also allow to separately configure memory and other configurations available to [aws-lambda](https://github.com/serverless-components/aws-lambda):

```yaml
myNextApp:
  component: serverless-next.js
  inputs:
    bucketName: my-bucket
    lambdas:
      default:
        memory: 1024
      api:
        routes: [/api/v3]
        memory: 512
      legacy:
        routes: [/api/v1, /api/v2]
        memory: 128
      ...
```

I would suggest defaulting to 1 lambda though

**Comment by sheerun** [comment]
Thinking about this probably prefix should enough for MVP:

```yaml
myNextApp:
  component: serverless-next.js
  inputs:
    bucketName: my-bucket
    lambdas:
      default:
        memory: 1024
      api:
        prefix: /api
        memory: 512
      blog:
        prefix: /blog 
        memory: 128
      ...
```

I can try to prepare PR. Are you OK with this config?

**Comment by boxabirds** [comment]
Note that AWS is trying to address the cold-start problem with [provisioned concurrency](https://aws.amazon.com/blogs/compute/new-for-aws-lambda-predictable-start-up-times-with-provisioned-concurrency/)

> "When enabled, Provisioned Concurrency is designed to keep your functions initialized and hyper-ready to respond in double-digit milliseconds at the scale you need. This new feature provides a reliable way to keep functions ready to respond to requests, giving you more precise control over the performance characteristics of your serverless applications. Builders can now choose the concurrency level for each Lambda function version or alias, including when and for how long these levels are in effect."

**Comment by barrysteyn** [comment]
@boxabirds This is amazing, thanks. However, for this to work properly, I would think that the underlying lambda component would need to be updated to take advantage of this new feature. While I don't think this would be a big problem, the Serverless component maintainers don't seem to be interested (scary), and so when (or if) they eventually become serious about taking PRs, then this could happen.

**Comment by boxabirds** [comment]
Cheers. Looks like they added it last month https://github.com/serverless/serverless/pull/7043

**Comment by barrysteyn** [comment]
@boxabirds Unfortunately, it needs to be added to [@serverless/aws-lambda](https://github.com/serverless-components/aws-lambda). This repo does not use the old version of serverless, it only uses serverless components.

**Comment by danielcondemarin** [comment]
Closing this, for cold starts support for provisioned concurrency will be added as per https://aws.amazon.com/about-aws/whats-new/2019/12/aws-lambda-announces-provisioned-concurrency/. I'll raise a separate issue for that.

**Comment by wis** [comment]
> Provisioned Concurrency is not supported with Lambda@Edge.

source: https://aws.amazon.com/blogs/aws/new-provisioned-concurrency-for-lambda-functions/

**Comment by lukemarshdv** [comment]
@danielcondemarin there's been chat about adding vpc configuration to this and it looks like the aws-lambda package now supports it. The question is - will it work?

**Comment by medv** [comment]
Lambda@Edge will not allow VPC or provisioned concurrency - if you consider that they live at 160+ edge locations then VPC not being allowed makes sense, and provisioned concurrency set to 1 would give you something like a $700 bill if it was allowed.

**Comment by wis** [comment]
@medv true.
cold start time is the biggest issue with Lambda@Edge rendering, edge locations being geographically closer to the client saves from 100ms to 300ms (due  to ([in part](https://royal.pingdom.com/theoretical-vs-real-world-speed-limit-of-ping/)) speed of light in fiber-optics) but the cold start time is easily double that. 
I currently use this component, which I think is awesome itself*, but I've always been looking for a way to deploy next.js serverless to Cloudflare Workers (which has more edge locations than Lambda@Edge, does not suffer from cold start times, but is not a node.js environment), I found about [FAB](https://github.com/fab-spec/fab) few months ago, It had next js deployment to CF workers as one of it's goals, but it was not ready, I just checked it out again it seems that it is now. 
 
* only because it makes deployment and scaling (at least the rendering part) easy.
